core/src/main/java/org/pentaho/di/core/row/value/ValueMetaBase.java

```diff
@@ -0,0 +1,5534 @@
+  }
+
+  protected Long readInteger( DataInputStream inputStream ) throws IOException {
+    Long l = new Long( inputStream.readLong() );
+    // System.out.println("Read integer("+getName()+") ["+l+"]");
+    return l;
+  }
+
+  protected void writeInteger( DataOutputStream outputStream, Integer number ) throws IOException {
+    outputStream.writeInt( number.intValue() );
+  }
+
+  protected Integer readSmallInteger( DataInputStream inputStream ) throws IOException {
+    Integer i = Integer.valueOf( inputStream.readInt() );
+    // System.out.println("Read index integer("+getName()+") ["+i+"]");
+    return i;
+  }
+
+  protected void writeBinary( DataOutputStream outputStream, byte[] binary ) throws IOException {
+    outputStream.writeInt( binary.length );
+    outputStream.write( binary );
+  }
+
+  protected byte[] readBinary( DataInputStream inputStream ) throws IOException {
+    int size = inputStream.readInt();
+    byte[] buffer = new byte[size];
+    inputStream.readFully( buffer );
+
+    // System.out.println("Read binary("+getName()+") with size="+size);
+
+    return buffer;
+  }
+
+  @Override
+  public void writeMeta( DataOutputStream outputStream ) throws KettleFileException {
+    try {
+      int type = getType();
+
+      // Handle type
+      outputStream.writeInt( type );
+
+      // Handle storage type
+      outputStream.writeInt( storageType );
+
+      switch ( storageType ) {
+        case STORAGE_TYPE_INDEXED:
+          // Save the indexed strings...
+          if ( index == null ) {
+            outputStream.writeInt( -1 ); // null
+          } else {
+            outputStream.writeInt( index.length );
+            for ( int i = 0; i < index.length; i++ ) {
+              try {
+                switch ( type ) {
+                  case TYPE_STRING:
+                    writeString( outputStream, (String) index[i] );
+                    break;
+                  case TYPE_NUMBER:
+                    writeNumber( outputStream, (Double) index[i] );
+                    break;
+                  case TYPE_INTEGER:
+                    writeInteger( outputStream, (Long) index[i] );
+                    break;
+                  case TYPE_DATE:
+                    writeDate( outputStream, (Date) index[i] );
+                    break;
+                  case TYPE_BIGNUMBER:
+                    writeBigNumber( outputStream, (BigDecimal) index[i] );
+                    break;
+                  case TYPE_BOOLEAN:
+                    writeBoolean( outputStream, (Boolean) index[i] );
+                    break;
+                  case TYPE_BINARY:
+                    writeBinary( outputStream, (byte[]) index[i] );
+                    break;
+                  default:
+                    throw new KettleFileException( toString()
+                        + " : Unable to serialize indexe storage type for data type " + getType() );
+                }
+              } catch ( ClassCastException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)